contracts/l2/FuseController.sol

```diff
@@ -0,0 +1,309 @@
+
+        // Change the controller to the upgrade contract.
+        registry.setNode(
+            uint256(node),
+            _pack(
+                address(upgradeContract),
+                owner,
+                resolver,
+                expiry,
+                fuses,
+                renewalController
+            )
+        );
+
+        // Call the new contract to notify it of the upgrade.
+        upgradeContract.upgradeFrom(node, extraData);
+    }
+
+    /*******************
+     * Node Owner functions *
+     *******************/
+
+    function setResolver(uint256 id, address newResolver) external {
+        // get tokenData
+        bytes memory tokenData = registry.getData(id);
+        (
+            address owner,
+            ,
+            uint64 expiry,
+            uint96 fuses,
+            address renewalController
+        ) = _unpack(tokenData);
+        bool isAuthorized = registry.getAuthorization(id, owner, msg.sender);
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(bytes32(id), msg.sender);
+        }
+
+        registry.setNode(
+            id,
+            _pack(
+                address(this),
+                owner,
+                newResolver,
+                expiry,
+                fuses,
+                renewalController
+            )
+        );
+    }
+
+    function setSubnode(
+        bytes32 node,
+        uint256 label,
+        address subnodeOwner,
+        address subnodeResolver,
+        uint64 subnodeExpiry,
+        uint96 subnodeFuses,
+        address subnodeRenewalController
+    ) external {
+        bytes memory tokenData = registry.getData(uint256(node));
+        (address owner, , , , ) = _unpack(tokenData);
+        bool isAuthorized = registry.getAuthorization(
+            uint256(node),
+            owner,
+            msg.sender
+        );
+
+        if (owner != msg.sender && !isAuthorized) {
+            revert Unauthorised(node, msg.sender);
+        }
+
+        registry.setSubnode(
+            uint256(node),
+            label,
+            _pack(
+                address(this),
+                subnodeOwner,
+                subnodeResolver,
+                subnodeExpiry,
+                subnodeFuses,
+                subnodeRenewalController
+            ),
+            msg.sender,
+            subnodeOwner
+        );
+    }
+
+    /*******************
+     * Owner only functions *
+     *******************/
+
+    // A function that sets the upgrade contract.
+    function setUpgradeController(
+        IControllerUpgrade _upgradeContract
+    ) external onlyOwner {
+        upgradeContract = _upgradeContract;
+    }
+
+    /**********************
+     * Internal functions *
+     **********************/
+
+    function _isExpired(bytes memory tokenData) internal view returns (bool) {
+        (, , uint64 expiry, , ) = _unpack(tokenData);
+        return expiry <= block.timestamp;
+    }
+
+    function _unpack(
```

Comment:
As I mentioned before - just use `calldata` instead of `memory` and you don't need assembly.

Reply:
Using `memory` allows for the function `ownerOfWithData`, which saves an external call.

Reply:
Why? `ownerOfWithData` can specify that its argument is `calldata` too, and pass it straight to `_unpack`.

Reply:
For example in this function in the controller:

`   function ownerOf(bytes32 node) external view returns (address) {
        //get the tokenData
        bytes memory tokenData = registry.getData(uint256(node));
        (address owner, , , , ) = _unpack(tokenData);
        return owner;
    }`
    
    I just get the data from the registry and then call _unpack, but tokenData in this case is memory and not calldata.